node/narwhal/tests/common/primary.rs

```diff
@@ -12,23 +12,193 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-use crate::common::{CurrentNetwork, MockLedgerService};
+use crate::common::{
+    utils::{fire_unconfirmed_solutions, fire_unconfirmed_transactions, initialize_logger},
+    CurrentNetwork,
+    MockLedgerService,
+};
 use snarkos_account::Account;
 use snarkos_node_narwhal::{
     helpers::{init_primary_channels, Committee, PrimarySender, Storage},
     Primary,
+    MAX_BATCH_DELAY,
     MAX_GC_ROUNDS,
     MIN_STAKE,
 };
 use snarkvm::prelude::TestRng;
 
+use std::{
+    collections::HashMap,
+    ops::{Deref, DerefMut},
+    sync::Arc,
+    time::Duration,
+};
+
 use indexmap::IndexMap;
 use itertools::Itertools;
-use std::collections::HashMap;
+use parking_lot::RwLock;
+use tokio::{task::JoinHandle, time::sleep};
 use tracing::*;
 
+#[derive(Clone)]
+pub struct TestNetworkConfig {
+    pub num_nodes: u16,
+    pub connect_all: bool,
+    pub fire_cannons: bool,
+    pub log_level: Option<u8>,
+    pub log_connections: bool,
+}
+
+#[derive(Clone)]
+pub struct TestNetwork {
+    pub config: TestNetworkConfig,
+    pub primaries: HashMap<u16, TestPrimary>,
+}
+
+#[derive(Clone)]
+pub struct TestPrimary {
+    pub id: u16,
+    pub primary: Primary<CurrentNetwork>,
+    pub sender: Option<PrimarySender<CurrentNetwork>>,
+    pub handles: Arc<RwLock<Vec<JoinHandle<()>>>>,
```

Comment:
```suggestion
    pub handles: Arc<Mutex<Vec<JoinHandle<()>>>>,
```
these won't be read that much